File: solutions/0056-merge-intervals/solution.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        vector<vector<int>> ans;
        sort(intervals.begin(),intervals.end());
        ans.push_back(intervals[0]);
        for(int i=1;i<intervals.size();i++){
            int a = ans[ans.size()-1][0];
            int b = ans[ans.size()-1][1];
            int c = intervals[i][0];
            int d = intervals[i][1];
            if(a<=c and c<=b){
                if(d<=b){
                    
                }
                else{
                    ans[ans.size()-1][1]=d;
                }
            }
            else{
                ans.push_back({c,d});
            }
        }
        return ans;
    }
};

File: solutions/0056-merge-intervals/solution.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort()
        ans = []
        ans.append(intervals[0])
        prev = intervals[0]
        for i in range(1,len(intervals)):
            curr = intervals[i]
            prev = ans[-1]
            # compare two
            # second intervals will start after first (gurantee)
            # second interval starts after first ends
            if curr[0] > prev[1]:
                # no need to merge
                ans.append(curr)
            else:
                # second interval ends after frist ends
                ans[-1] = [prev[0],max(curr[1],prev[1])]
        return ans


        
